modules/core/include/opencv2/core/directx.hpp

```diff
@@ -127,7 +127,21 @@ CV_EXPORTS void convertToD3D11Texture2D(InputArray src, ID3D11Texture2D* pD3D11T
 //
 //! @param pD3D11Texture2D - source D3D11 texture
 //! @param dst             - destination OutputArray
-CV_EXPORTS void convertFromD3D11Texture2D(ID3D11Texture2D* pD3D11Texture2D, OutputArray dst);
+//! @param device - pointer to cv::ocl::Device that will be used for context creation
+CV_EXPORTS void convertFromD3D11Texture2D(ID3D11Texture2D* pD3D11Texture2D, OutputArray dst, cv::ocl::Device* device = nullptr);
```

Comment:
> cv::ocl::Device* device
> device - pointer to cv::ocl::Device that will be used for context creation

Please reuse `OpenCLExecutionContext` as we don't pass OpenCL device/context through API calls explicitly.

---

D3D11 compatible OpenCL device/context must be initialized before this call.

Reply:
I create OpenCLExecutionContext by `cl_device_id` from `cv::ocl::Device`.
How I can reuse context? Where I can get it?
And
What if I want to choose a device instead of working with automatically selected device?

Reply:
Take a look on `initializeContextFromD3D11Device` functions.

Reply:
I use the same approach. But I added device selection for user. What if I can sacrifice performance and transfer the task to another device (not preferred)?

Can I use configuration parameters in API calls or create custom function for this?

Reply:
if we need to work with another device, then we should switch `OpenCLExecutionContext` instead of passing extra "device" parameters.

Reply:
Ok. Can `OpenCLExecutionContext` be a parameter of function?
More context: I do not want to search for a device for each texture2D to UMat conversion, for example, for video. I can find preferred device(ExecContext) once and reuse it for all frames.

Reply:
> Can OpenCLExecutionContext be a parameter of function?

No.
It is a **thread-local** "proxy" for OpenCL device/context.

Reply:
Is there an example of switching `OpenCLExecutionContext`?

Reply:
When I should use `OpenCLExecutionContext::create` and `OpenCLExecutionContext::getCurrent()`?
How I can create ExecContext and reuse it?

Reply:
Again, take a look on `initializeContextFromD3D11Device()` functions.
Why you don't use them?

Reply:
I don't want to search the same cl_device_id for every pD3D11Device of Texture2D (frame of video) (clGetDeviceIDsFromD3D11KHR).
I chose way when I create ExecContext for one device. Found a working solution (Ok. It is wrong. I understood).

Why do not I use `convertFromD3D11Texture2D` (based on `OpenCLExecutionContext::getCurrent()`)?
- because it doesn't work: "OpenCL: Context initialized without DirectX interoperability" (I have cl_khr_d3d11_sharing). `getUserContext` returns empty implementation, looks like `setUserContext` is missed. MB I must call something before `convertFromD3D11Texture2D`?

I'm completely confused with the ExecutionContext.

Reply:
Ok. I think that understood behavior.
I call `initializeContextFromD3D11Device` once and after that I call `convertFromD3D11Texture2D` for the rest frames of video.
@alalek Is it correct behavior?
And what if I will call `convertFromD3D11Texture2D` in different threads? Will it broke logic or `ExecutionContext::getCurrent()` will reuse context from `initializeContextFromD3D11Device`'s thread (`main thread (threadID = 0)`)?